### PR TITLE
fix(cuda): correct DeepSeek-V4 forward, RoPE/YaRN, stops, and determinism

### PR DESCRIPTION
**Pull request by zyma98**
The following fixes in this PR make DS4 work on CUDA. Each commit fixes one problem as summarized below:

- **Hash-layer MoE weights** — still select experts from `tid2eid`, but weight them with renormalized `sqrtsoftplus` router probs × `routed_scaling_factor` instead of uniform 1/K.
- **SwiGLU clamp** — apply `swiglu_limit=10` pre-activation clamps on routed and shared experts.
- **RoPE pairing** — rotate adjacent (GPT-J) dims; Pie had been using NeoX half/half.
- **Compressed-layer YaRN** — use `compress_rope_theta` with YaRN (`freq_scale=1/factor`) on C4/C128 layers for forward, inverse, and compressed-KV RoPE.
- **Sliding window** — raw attention honors `sliding_window` (128); longer context only via compressed attention.
- **EOS / stop tokens** — skip HF `added_tokens` during GPT-2 vocab re-key and register them as literal UTF-8 so `token_to_id` resolves stop strings.
- **HC RMSNorm** — replace cross-warp `atomicAdd` with a fixed-order reduction for run-to-run determinism.
- **CLI** — map HF `deepseek_v3` / `deepseek_v4` in `pie model download`’s arch table.

## Summary by CodeRabbit

- **New Features**
  - Added support for DeepSeek V3 and V4 model architectures.
  - Improved long-context RoPE behavior with YaRN interpolation.
  - Added clamped SwiGLU activation support for more stable inference.
  - Added router-logit-based expert weighting for MoE models.
  - Improved tokenizer handling for added and special tokens.

- **Improvements**
  - Made RMS normalization results deterministic across runs.
  - Updated sliding-window attention behavior for configured windows.

### CODE REVIEW

File: driver/cuda/src/kernels/dsv4_routing.cu [hunk 1]
```diff
@@ -68,28 +68,45 @@ __global__ void topk_sqrtsoftplus_kernel(
...
+    float sum = 0.f;
     for (int k = 0; k < K; ++k) {
-        out_idx[k] = static_cast<std::int32_t>(row[k]);
-        out_w[k] = weight_per_expert;
+        const int e = static_cast<int>(row[k]);
+        out_idx[k] = e;
+        const float p = (e >= 0 && e < E)
+            ? sqrtsoftplus(__bfloat162float(logits[e]))
+            : 0.f;
+        out_w[k] = p;
```
[the remark this answers is not in the document]
Comment (zyma98) [reply driver/cuda/src/kernels/dsv4_routing.cu:103]: The model weights are considered trusted for now. No need to guard against out-of-bound expert ID.

File: driver/cuda/src/kernels/dsv4_routing.cu [hunk 2]
```diff
@@ -117,20 +134,24 @@ void launch_topk_sqrtsoftplus_bf16(
...
     int tokens,
     int vocab_size,
+    int num_experts,
     int top_k,
-    float weight_per_expert,
+    float routed_scaling_factor,
     cudaStream_t stream)
 {
-    if (tokens <= 0 || top_k <= 0) return;
+    if (tokens <= 0 || num_experts <= 0 || top_k <= 0) return;
```
[the remark this answers is not in the document]
Comment (zyma98) [reply driver/cuda/src/kernels/dsv4_routing.cu:147]: he model weights are considered trusted for now. No need to consider zero or negative expert number case.
